### PR TITLE
BFD-2254: mdc resources returned count fix

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-2254](https://jira.cms.gov/browse/BFD-2254)

**User Story or Bug Summary:**

Bug description:

All FHIR requests are supposed to set the `resources_returned_count` in the MDC to the number of resources returned in the response. However, the EOB calls have some cases where zero is returned even though there are in fact more than zero resources returned in the payload.

Specifically, (using V2 as an example) the number of resources is being set in the MDC [here](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java#L1567) using the size of the bundle but the bundle size is not itself set until later on [here](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java#L1588).

AC:

The issue described above should be corrected to ensure that the EOB endpoint (v1 and v2) always sets the correct value for resources_returned_count
A BFD Insights query running in PROD should be used to perform a validation by ensuring that starting from the application of the fix and going forward for at least 24 hours, that all EOB calls with resources_returned_count set to zero have an output size <= the largest possible empty EOB response (currently believed to be around 350 bytes). See notes for such a query.
BFD insights [documentation](https://github.com/CMSgov/beneficiary-fhir-data/blob/master/insights/terraform/projects/bfd/README.md) should be updated to note the window during which the data was incorrect.
Notes:

This should return zero when the defect is fixed (replace the timestamp with a timestamp after the fix is applied):

---

### What Does This PR Do?

Fixes the mdc resources returned count

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java [hunk 1]
```diff
@@ -1586,6 +1581,10 @@ public static Bundle createBundle(
                 ? Date.from(transactionTime)
                 : Date.from(maxBundleDate));
     bundle.setTotal(resources.size());
+
+    // Add number of resources to MDC logs
+    LoggingUtils.logResourceCountToMdc(bundle.getTotal());
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java:1586]: This is going to be incorrect when there are more resources than a page holds and there should be a test for that case.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java [hunk 2]
```diff
@@ -641,4 +663,92 @@ public void addCareTeamMemberWithNpiOrgShouldCreateCareTeamEntry() {
...
+                FdaDrugCodeDisplayLookup.createDrugCodeLookupForTesting(),
+                NPIOrgLookup.createNpiOrgLookupForTesting()),
+            claim);
+    IParser parser = fhirContext.newJsonParser();
+    String json = parser.encodeResourceToString(genEob);
+    List<IBaseResource> eobs = new ArrayList<IBaseResource>();
+    eobs.add(parser.parseResource(ExplanationOfBenefit.class, json));
+
+    Bundle bundle = TransformerUtilsV2.createBundle(paging, eobs, Instant.now());
+    assertEquals(1, bundle.getTotal());
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java:705]: Might want to make total return 2-3 if its not a big lift, just to ensure this works with many (using 1 doesnt check that multiple are counted correctly)

Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java:705]: Should have an assert for the MDC value

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtilsTest.java [hunk 3]
```diff
@@ -266,6 +289,94 @@ public void createReferenceForIdentifierTypeTest() {
...
+    eobs.add(parser.parseResource(ExplanationOfBenefit.class, json));
+
+    Bundle bundle = TransformerUtils.createBundle(paging, eobs, Instant.now());
+    assertEquals(1, bundle.getTotal());
+  }
+
+  /**
+   * Verifies that {@link gov.cms.bfd.server.war.stu3.providers.TransformerUtils#createBundle
+   * (OffsetLinkBuilder paging, List<IBaseResource> resources, Instant transactionTime)} sets bundle
+   * with paging size correctly} correctly.
```
Comment (mackec1) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/TransformerUtilsTest.java:336]: minor nit...comment seems a bit whacked.

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java [hunk 4]
```diff
@@ -641,4 +663,92 @@ public void addCareTeamMemberWithNpiOrgShouldCreateCareTeamEntry() {
...
+   * (OffsetLinkBuilder paging, List<IBaseResource> resources, Instant transactionTime)} sets bundle
+   * with paging size correctly} correctly.
+   */
+  @Test
+  public void createBundleWithPaging() throws IOException {
+
+    RequestDetails requestDetails = mock(RequestDetails.class);
+    Map<String, String[]> pagingParams = new HashMap<String, String[]>();
+    pagingParams.put(Constants.PARAM_COUNT, new String[] {"1"});
+    pagingParams.put("startIndex", new String[] {"1"});
```
Comment (lsmitchell) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java:719]: Is this only making one page? Shouldnt there be multiple pages to properly test this works with paging?

**Review by lsmitchell — COMMENTED** [review]
Minor questions/comments about the tests.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java (hunk 1 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2.java]: Moved

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java (hunk 2 above)
Comment (cbrunefearless) [reply apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/r4/providers/TransformerUtilsV2Test.java]: Done

**Review by lsmitchell — APPROVED** [review]
Thanks for adding additional testing!

**Review by christrotteradhoc — APPROVED** [review]
Thanks for getting this one turned around quickly Chris!

**Pull request merged**